### PR TITLE
csi: add a new flag to disable csi driver

### PR DESCRIPTION
**Pull request by parth-gr**
added a new flag ROOK_CSI_DISABLE_DRIVER
to disable csi controller.

Signed-off-by: parth-gr <partharora1010@gmail.com>

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/operator.yaml [hunk 1]
```diff
@@ -35,6 +35,8 @@ data:
   ROOK_CSI_ENABLE_RBD: "true"
   # Enable the CSI NFS driver. To start another version of the CSI driver, see image properties below.
   ROOK_CSI_ENABLE_NFS: "false"
+  # Enable the CSI driver.
+  ROOK_CSI_DISABLE_CSI_DRIVER: "false"
```
Comment (Madhu-1) [inline deploy/examples/operator.yaml:39]: `CSI_DISABLE_DRIVER` or `ROOK_CSI_DISABLE_DRIVER` might be the better name and comment.
1. Update the comment to says set this to diable csi driver
2. add the same to operator-openshift.yaml and helm charts

File: pkg/operator/ceph/csi/controller.go [hunk 2]
```diff
@@ -133,10 +133,44 @@ func (r *ReconcileCSI) Reconcile(context context.Context, request reconcile.Requ
 // allow overriding for unit tests
 var reconcileSaveCSIDriverOptions = SaveCSIDriverOptions
 
+func (r *ReconcileCSI) csiDriverDisabled() bool {
+	DisableCSI, err := strconv.ParseBool(k8sutil.GetValue(r.opConfig.Parameters, "ROOK_CSI_DISABLE_CSI_DRIVER", "false"))
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/controller.go:137]: use small alphabet for variable name

File: pkg/operator/ceph/csi/controller.go [hunk 3]
```diff
@@ -133,10 +133,44 @@ func (r *ReconcileCSI) Reconcile(context context.Context, request reconcile.Requ
 // allow overriding for unit tests
 var reconcileSaveCSIDriverOptions = SaveCSIDriverOptions
 
+func (r *ReconcileCSI) csiDriverDisabled() bool {
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/controller.go:136]: we dont really need helper function to get one value, please directly parse from the configmap when required, the same way we are doing for other variables in reconcile function

**Review by Madhu-1 — CHANGES_REQUESTED** [review]

**Comment by Madhu-1** [comment]
@parth-gr This should be a new PR, don't cherry-pick a commit from unmerged PR's

File: deploy/examples/operator.yaml (hunk 1 above)
Comment (parth-gr) [reply deploy/examples/operator.yaml]: Updated

**Comment by parth-gr** [comment]
> @parth-gr This should be a new PR, don't cherry-pick a commit from unmerged PR's

Yaa its a new commit added

File: pkg/operator/ceph/csi/controller.go [hunk 4]
```diff
@@ -137,6 +137,33 @@ func (r *ReconcileCSI) reconcile(request reconcile.Request) (reconcile.Result, e
...
+		// Populate the operator's config
+		r.opConfig.Parameters = opConfig.Data
+	}
+
+	// do not recocnile if csi driver is disabled
+	disableCSI, err := strconv.ParseBool(k8sutil.GetValue(r.opConfig.Parameters, "CSI_DISABLE_CSI_DRIVER", "false"))
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "unable to parse value for 'CSI_DISABLE_CSI_DRIVER")
+	} else if disableCSI {
+		return reconcile.Result{}, nil
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/controller.go:164]: Add a log to convey that csi driver is disabled?

**Review by Madhu-1 — COMMENTED** [review]
LGTM, please remove extra sign-off-by in your comment message and also update commit message for new flag

File: deploy/examples/operator.yaml (hunk 1 above)
Comment (parth-gr) [reply deploy/examples/operator.yaml]: How about `DISABLE_CSI_DRIVER`? Currently used `ROOK_CSI_DISABLE_DRIVER`

Comment (Madhu-1) [reply deploy/examples/operator.yaml]: trying to keep `ROOK` or `CSI` as starting key. `ROOK_CSI_DISABLE_DRIVER` is also fine

File: pkg/operator/ceph/csi/controller.go [hunk 5]
```diff
@@ -137,6 +137,34 @@ func (r *ReconcileCSI) reconcile(request reconcile.Request) (reconcile.Result, e
...
+		// Populate the operator's config
+		r.opConfig.Parameters = opConfig.Data
+	}
+
+	// do not recocnile if csi driver is disabled
+	disableCSI, err := strconv.ParseBool(k8sutil.GetValue(r.opConfig.Parameters, "ROOK_CSI_DISABLE_DRIVER", "false"))
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "unable to parse value for 'ROOK_CSI_DISABLE_DRIVER")
+	} else if disableCSI {
+		logger.Debug("ceph csi driver is disabled")
```
Comment (subhamkrai) [inline pkg/operator/ceph/csi/controller.go:164]:
```suggestion
		logger.Info("ceph csi driver is disabled")
```

this should be info?

**Review by subhamkrai — CHANGES_REQUESTED** [review]
Question:
what will happen if someone sets this true on the running cluster?

And how's testing looking for this?

**Comment by parth-gr** [comment]
> Question: what will happen if someone sets this true on the running cluster?
> 
> And how's testing looking for this?

The csi pods keep on running, but if they went in bad state there is no controller managing them

**Comment by leelavg** [comment]
> ROOK_CSI_DISABLE_DRIVER
- Maybe it should be the opposite `ROOK_CSI_ENABLE_DRIVER` to conform to existing naming conventions like `ROOK_CSI_ENABLE_CEPHFS` and `ROOK_CSI_ENABLE_RBD` 🤔

File: deploy/examples/operator-openshift.yaml [hunk 6]
```diff
@@ -121,6 +121,8 @@ data:
   ROOK_CSI_ENABLE_RBD: "true"
   # Enable the CSI NFS driver. To start another version of the CSI driver, see image properties below.
   ROOK_CSI_ENABLE_NFS: "false"
+  # Disable the CSI driver.
+  ROOK_CSI_DISABLE_DRIVER: "false"
```
Comment (subhamkrai) [inline deploy/examples/operator-openshift.yaml:125]:
```suggestion
  ROOK_CSI_DISABLE_DRIVER: "false"
```

@parth-gr JFYI, when backporting to downstream you will require to add this flag in operator env and run `make gen-csv`

File: pkg/operator/ceph/csi/controller.go [hunk 7]
```diff
@@ -137,6 +137,34 @@ func (r *ReconcileCSI) reconcile(request reconcile.Request) (reconcile.Result, e
...
+			logger.Debug("operator's configmap resource not found. will use default value or env var.")
+			r.opConfig.Parameters = make(map[string]string)
+		} else {
+			// Error reading the object - requeue the request.
+			return opcontroller.ImmediateRetryResult, errors.Wrap(err, "failed to get operator's configmap")
+		}
+	} else {
+		// Populate the operator's config
+		r.opConfig.Parameters = opConfig.Data
+	}
```
Comment (subhamkrai) [inline pkg/operator/ceph/csi/controller.go:157]:
```suggestion
	opNamespaceName := types.NamespacedName{Name: opcontroller.OperatorSettingConfigMapName, Namespace: r.opConfig.OperatorNamespace}
	opConfig := &v1.ConfigMap{}
	err := r.client.Get(r.opManagerContext, opNamespaceName, opConfig)
	if err != nil {
		if kerrors.IsNotFound(err) {
			logger.Debug("operator's configmap resource not found. will use default value or env var.")
			r.opConfig.Parameters = make(map[string]string)
		} else {
			// Error reading the object - requeue the request.
			return opcontroller.ImmediateRetryResult, errors.Wrap(err, "failed to get operator's configmap")
		}
	} else {
		// Populate the operator's config
		r.opConfig.Parameters = opConfig.Data
	}
```
we moved this block of code here to get the `r.opConfig.Parameters` values?

File: deploy/examples/operator-openshift.yaml (hunk 6 above)
Comment (subhamkrai) [reply deploy/examples/operator-openshift.yaml:125]: never mind I see https://github.com/red-hat-storage/rook/pull/600

File: pkg/operator/ceph/csi/controller.go (hunk 7 above)
Comment (Madhu-1) [reply pkg/operator/ceph/csi/controller.go:157]: Yes that is correct, we need to read from the configmap

**Comment by parth-gr** [comment]
> > ROOK_CSI_DISABLE_DRIVER
> 
> * Maybe it should be the opposite `ROOK_CSI_ENABLE_DRIVER` to conform to existing naming conventions like `ROOK_CSI_ENABLE_CEPHFS` and `ROOK_CSI_ENABLE_RBD` 🤔

The operation disables the CSI controller so let's keep the keyword disabled. And if we keep enable, then we need to check `!enable` in the if condition, which is a bad syntax.

File: deploy/examples/operator-openshift.yaml (hunk 6 above)
Comment (parth-gr) [reply deploy/examples/operator-openshift.yaml:125]: Btw a different discussion, we generally do auto branch syncup upstream to downstream, so how do you think we need a way to auto run the `make gen-csv` cmd if cahgnes are specific to csv?

Comment (parth-gr) [reply deploy/examples/operator-openshift.yaml:125]: @subhamkrai ^^

**Review by subhamkrai — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]

File: deploy/examples/operator-openshift.yaml (hunk 6 above)
Comment (subhamkrai) [reply deploy/examples/operator-openshift.yaml:125]: after FF it will not be auto sync

Comment (parth-gr) [reply deploy/examples/operator-openshift.yaml:125]: but what about PRs before ff?

Comment (subhamkrai) [reply deploy/examples/operator-openshift.yaml:125]: We'll require to resolve manually

**Pull request merged**